From 9aaffb8bddd3c4804d7ee3489437749018ed91b3 Mon Sep 17 00:00:00 2001
From: pdiopX <piotrx.diop@intel.com>
Date: Tue, 2 Jul 2013 10:50:42 +0200
Subject: [PORT FROM MAIN] [AUDIO] Make distinction between stopped and not-yet-started input

BZ: 115755

There was a race condition when a client calls getInput(), but is preempted
by a second client getInput() call for the same source. In that case, after
the release of the latter, the input is given back to the first client if he
is still alive (see patch I57b4b27151475d449d503d8e12d7d57159d04d77).
In this situation the startInput() call from the first client will fail as
the policy assumes that it has already started and is active.

This fix implements a new inputDescriptor parameter, mHasStarted to distinguish
an input that has been stopped (and therefore already called startInput()) from
an input that did not yet start.

Change-Id: Id4a57e8c13a5061338be3754075fbd5ae00ec5c3
Signed-off-by: pdiopX <piotrx.diop@intel.com>
Reviewed-on: http://android.intel.com:8080/116551
Reviewed-by: Di Folco, Neil <neil.di.folco@intel.com>
Reviewed-by: Gonzalve, Sebastien <sebastien.gonzalve@intel.com>
Reviewed-by: Gaffie, Francois <francois.gaffie@intel.com>
Tested-by: Dixon, CharlesX <charlesx.dixon@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 audio/AudioPolicyManagerBase.cpp                 |    2 +-
 include/hardware_legacy/AudioPolicyManagerBase.h |    1 +
 2 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index 5d3863c..fb6aed4 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -3463,7 +3463,7 @@ status_t AudioPolicyManagerBase::AudioOutputDescriptor::dump(int fd)
 AudioPolicyManagerBase::AudioInputDescriptor::AudioInputDescriptor(const IOProfile *profile)
     : mSamplingRate(0), mFormat((audio_format_t)0), mChannelMask((audio_channel_mask_t)0),
       mDevice(AUDIO_DEVICE_NONE), mRefCount(0),
-      mInputSource(0), mProfile(profile)
+      mInputSource(0), mProfile(profile), mHasStarted(0)
 {
 }
 
diff --git a/include/hardware_legacy/AudioPolicyManagerBase.h b/include/hardware_legacy/AudioPolicyManagerBase.h
index 7ab48f5..0deae36 100644
--- a/include/hardware_legacy/AudioPolicyManagerBase.h
+++ b/include/hardware_legacy/AudioPolicyManagerBase.h
@@ -298,6 +298,7 @@ protected:
             uint32_t mRefCount;                         // number of AudioRecord clients using this output
             int      mInputSource;                      // input source selected by application (mediarecorder.h)
             const IOProfile *mProfile;                  // I/O profile this output derives from
+            bool     mHasStarted;                       // to indicate that the audiorecord has started
         };
 
         // stream descriptor used for volume control
-- 
1.7.4.1

